File: src/main/java/202.happy-number.java
import java.util.HashSet;

class Q202 {

  public boolean isHappy(int n) {
    return isHappy(n, new HashSet<>());
  }

  private boolean isHappy(int n, HashSet<Integer> set) {
    int sum = 0;
    while (n != 0) {
      int x = n % 10;
      sum += x * x;
      n = n / 10;
    }
    // 找到 -> 跳出循环
    if (sum == 1) {
      return true;
    }
    // 重复了说明死循环 -> 跳出循环
    if (set.contains(sum)) {
      return false;
    }
    set.add(sum);
    return isHappy(sum, set);
  }
}

File: src/main/scala/202.happy-number.scala
object Q202Scala {
  def isHappy(n: Int): Boolean = {
    @scala.annotation.tailrec
    def loop(n: Int, set: Set[Int]): Boolean = n match {
      case 1                    => true
      case s if set.contains(s) => false
      case s =>
        @scala.annotation.tailrec
        def sum(n: Int, acc: Int): Int =
          if (n == 0) acc
          else {
            val x = n % 10
            sum(n / 10, acc + x * x)
          }
        loop(sum(s, 0), set + s)
    }
    loop(n, Set())
  }
}
